src/ast/semantic_analyser.cpp

```diff
@@ -1385,30 +1385,44 @@ void SemanticAnalyser::visit(ArrayAccess &arr)
   SizedType &indextype = arr.indexpr->type;
 
   if (is_final_pass()) {
-    if (!type.IsArrayTy())
+    if (!type.IsArrayTy() && !type.IsPtrTy())
     {
-      LOG(ERROR, arr.loc, err_)
-          << "The array index operator [] can only be used on arrays, found "
-          << type.type << ".";
+      LOG(ERROR, arr.loc, err_) << "The array index operator [] can only be "
+                                   "used on arrays and pointers, found "
+                                << type.type << ".";
       return;
     }
 
+    if (type.IsPtrTy() && type.GetPointeeTy()->GetSize() == 0)
+    {
+      LOG(ERROR, arr.loc, err_) << "The array index operator [] cannot be used "
+                                   "on a pointer to an unsized type (void *).";
+    }
+
     if (indextype.IsIntTy() && arr.indexpr->is_literal)
     {
-      Integer *index = static_cast<Integer *>(arr.indexpr);
+      if (type.IsArrayTy())
+      {
+        Integer *index = static_cast<Integer *>(arr.indexpr);
 
-      if ((size_t)index->n >= type.GetNumElements())
-        LOG(ERROR, arr.loc, err_)
-            << "the index " << index->n
-            << " is out of bounds for array of size " << type.GetNumElements();
+        if ((size_t)index->n >= type.GetNumElements())
+          LOG(ERROR, arr.loc, err_) << "the index " << index->n
+                                    << " is out of bounds for array of size "
+                                    << type.GetNumElements();
+      }
     }
     else {
       LOG(ERROR, arr.loc, err_) << "The array index operator [] only "
                                    "accepts literal integer indices.";
```

Comment:
I'm ok with this restriction for both constant pointers and arrays for now but perhaps in the future we can allow expressions as the array index. And if we happen to know both the size of the array and the index, we can warn if they're reading out of bounds.

For stuff like this I'd prefer to trust the user. But again, we can do it if someone asks later.

Reply:
Doing dynamic index would be awesome! I have a script where I loop through all cpus on the machine, and prints one line per cpu, and use the cpu index as the array (it is a pointer tho). This works awesome when removing this validation, but when sharing the "script" I have to concat probes in eg. bash and use one probe per cpu core.

Should I open an issue about this?

Reply:
But thank you for this @viktormalik!

btw. the smart git merge logic added the changelog to `v0.12`, but it is not part of that release... :sweat_smile:

Reply:
It should be part of the release. Can you file a bug report if it's not working?

Reply:
Hey @danobi!

The code works as a charm, so that is not a problem! The only way (currently) to use it is to compile master yourself. It is not a part of `0.21.0` [0] or `0.21.1`, even though the changelog in master says so [1].

If you look at the code in this PR, it is added to the `Unreleased` block, but when it is merged in b89f7611f5737eae [2], it is added to `0.21.0`, probably due to the git merging algorithm. :)


Again, thanks for your awesome work! :)

```
[0]: https://github.com/iovisor/bpftrace/commits/v0.12.1
[1]: https://github.com/iovisor/bpftrace/blob/master/CHANGELOG.md
[2]: https://github.com/iovisor/bpftrace/commit/b89f7611f5737eae37daefdecf949c3ea66b8464
```

Reply:
yes something went wonky there

```
$ git log --oneline upstream/master
...
118bf1e BTF: set correct size for struct/union types
b89f761 Support array indexing on pointers
373d43e Replace '@' prefix in map name with 'AT_' when creating map
d8d9ace (tag: v0.12.0, release) v0.12.0 release bump
```

Let's just do a .13 release in a few weeks, this one was looooong overdue

Reply:
@odinuge thanks for the feedback! Please open an issue for "expressions as array indices". Also, if you have a concrete usage example, add it to the issue so that we can evaluate the implementation. I'll try to get to that in near future.

Reply:
Will do @viktormalik! :smile: